xds/internal/client/client_xds.go

```diff
@@ -53,52 +58,137 @@ func UnmarshalListener(resources []*anypb.Any, logger *grpclog.PrefixLogger) (ma
 			return nil, fmt.Errorf("xds: failed to unmarshal resource in LDS response: %v", err)
 		}
 		logger.Infof("Resource with name: %v, type: %T, contains: %v", lis.GetName(), lis, lis)
-		routeName, err := getRouteConfigNameFromListener(lis, logger)
+
+		var (
+			lu  *ListenerUpdate
+			err error
+		)
+		if lis.GetApiListener() != nil {
+			lu, err = processClientSideListener(lis, logger)
+		} else {
+			lu, err = processServerSideListener(lis, logger)
+		}
 		if err != nil {
 			return nil, err
 		}
-		update[lis.GetName()] = ListenerUpdate{RouteConfigName: routeName}
+		update[lis.GetName()] = *lu
 	}
 	return update, nil
 }
 
-// getRouteConfigNameFromListener checks if the provided Listener proto meets
+// processClientSideListener checks if the provided Listener proto meets
 // the expected criteria. If so, it returns a non-empty routeConfigName.
-func getRouteConfigNameFromListener(lis *v3listenerpb.Listener, logger *grpclog.PrefixLogger) (string, error) {
+func processClientSideListener(lis *v3listenerpb.Listener, logger *grpclog.PrefixLogger) (*ListenerUpdate, error) {
```

Comment:
Maybe:

```go
func processListener(...) (*ListenerUpdater, error) {
	if lis.GetApiListener() != nil {
		return processClientSideListener(lis, logger)
	}
	processServerSideListener(lis, logger)
}
```

This simplifies the above code, and then below you can remove the check for `GetApiListener` in `processClientSideListener` since it's just a helper for this function.